### PR TITLE
feat/steps as date (fixes #811)

### LINKED ISSUE #811: Good First Issue: Allow `predict` method to accept date values as `steps`
Use branch 0.14.x as base.

**Summary**

Currently, the `steps` parameter in all Forecasters' `predict` methods only accepts an integer value. This integer defines how many observations to forecast into the future. We would like to extend this functionality so that `steps` can also accept a date (e.g., `'2020-01-01'`). If a date is provided, the function should calculate the appropriate number of observations corresponding to the time window between the last observation in the last window and the given date.

**Task**  
1. Create an auxiliary function, `_preprocess_steps_as_date(last_window: pd.Series, steps)` in the `utils` module:

- `last_window` is the last window of the series used to forecast the future. This is an argument of the `predict` method in all Forecasters.
- `steps` can be an integer or any datetime format that pandas allows to be passed to a `pd.DatetimeIndex` (e.g., string, pandas timestamp...).
- If the Forecaster was not fitted using a `pd.DatetimeIndex`, raise a `TypeError` with the message: "If the Forecaster was not fitted using a pd.DatetimeIndex, `steps` must be an integer."
- If the Forecaster was fitted using a `pd.DatetimeIndex`, this function will return the length of the time window between the last observation in the last window and the given date as an integer value.
- If the input `steps` is an integer, return the same integer.
- Create unit tests using pytest in the `utils.tests` folder.

```python
# Expected behavior
# ==============================================================================
last_window = pd.Series([1, 2, 3, 4, 5], index=pd.date_range('2020-01-01', periods=5, freq='D'))
_preprocess_steps_as_date(last_window, '2020-01-07') # expected output: 2

last_window = pd.Series([1, 2, 3, 4, 5], index=pd.date_range('2020-01-01', periods=5, freq='D'))
_preprocess_steps_as_date(last_window, 2) # expected output: 2

last_window = pd.Series([1, 2, 3, 4, 5], index=pd.RangeIndex(start=0, stop=5, step=1))
_preprocess_steps_as_date(last_window, '2020-01-07') # expected output: TypeError
```

2. Integrate this function in the `predict` method of the `ForecasterAutoreg` class.

**Acceptance Criteria**
- [ ] The `steps` parameter accepts both integer and date formats.
- [ ] The function correctly calculates the number of steps when a date is provided.
- [ ] Existing tests continue to pass.
- [ ] New test cases are added to verify the correct behavior for both int and date inputs.

**Full Example**

```python
# Expected behavior
# ==============================================================================
data = fetch_dataset(name="h2o", kwargs_read_csv={"names": ["y", "datetime"], "header": 0})

steps = 36
data_train = data[:-steps]
data_test  = data[-steps:]

forecaster = ForecasterAutoreg(
                 regressor = LGBMRegressor(random_state=123, verbose=-1),
                 lags      = 15 
             )
forecaster.fit(y=data_train['y'])

predictions = forecaster.predict(steps='2005-09-01') # As steps=3
```
2005-07-01    1.020833
2005-08-01    1.021721
2005-09-01    1.093488
Freq: MS, Name: pred, dtype: float64

### PR DESCRIPTION
**Pull request by imMoya**

### CODE REVIEW

File: skforecast/ForecasterAutoreg/ForecasterAutoreg.py [hunk 1]
```diff
@@ -1102,23 +1102,24 @@ def _binning_in_sample_residuals(
...
         """
         Create the inputs needed for the first iteration of the prediction 
         process. As this is a recursive process, the last window is updated at 
         each iteration of the prediction process.
         
         Parameters
         ----------
-        steps : int
-            Number of future steps predicted.
+        steps : Union[int, str, pd.Timestamp]
```
Comment (JavierEscobarOrtiz) [inline skforecast/ForecasterAutoreg/ForecasterAutoreg.py]: Adapt to follow the same format as other arguments, "steps : int, str, pandas Timestamp"

File: skforecast/ForecasterAutoreg/ForecasterAutoreg.py [hunk 2]
```diff
@@ -1412,8 +1415,9 @@ def predict(
...
```
Comment (JavierEscobarOrtiz) [inline skforecast/ForecasterAutoreg/ForecasterAutoreg.py]: Adapt to follow the same format as other arguments, "steps : int, str, pandas Timestamp"

File: skforecast/utils/utils.py [hunk 3]
```diff
@@ -1513,25 +1513,46 @@ def exog_to_direct_numpy(
...
     
     Parameters
     ----------
     index : pandas Index, None
         Original index.
-    steps : int
-        Number of steps to expand.
+    steps : Union[int, str, pd.Timestamp]
+        Number of steps to expand (if type is int) or
+        Datetime until the index is expanded (str formatted as YYYY-MM-DD or pd.Timestamp)
```
Comment (JavierEscobarOrtiz) [inline skforecast/utils/utils.py]: I will change the docstring to:

If `steps` is int, number of steps to expand. If str or pandas Datetime, expand index to this date (str must be convertible to pandas Datetime format).

File: skforecast/utils/utils.py [hunk 4]
```diff
@@ -1513,25 +1513,46 @@ def exog_to_direct_numpy(
...
     """
     Create a new index of length `steps` starting at the end of the index.
     
     Parameters
     ----------
     index : pandas Index, None
         Original index.
-    steps : int
-        Number of steps to expand.
+    steps : Union[int, str, pd.Timestamp]
```
Comment (JavierEscobarOrtiz) [inline skforecast/utils/utils.py]: Adapt to follow the same format as other arguments, "steps : int, str, pandas Timestamp"

File: skforecast/ForecasterAutoreg/ForecasterAutoreg.py [hunk 5]
```diff
@@ -1102,23 +1102,24 @@ def _binning_in_sample_residuals(
...
         process. As this is a recursive process, the last window is updated at 
         each iteration of the prediction process.
         
         Parameters
         ----------
-        steps : int
-            Number of future steps predicted.
+        steps : Union[int, str, pd.Timestamp]
+            Number of steps to expand (if type is int) or
+            Datetime until the index is expanded (str formatted as YYYY-MM-DD or pd.Timestamp).
```
Comment (JavierEscobarOrtiz) [inline skforecast/ForecasterAutoreg/ForecasterAutoreg.py]: In prediction methods, I will change the docstring to:

If `steps` is int, number of steps to predict. If str or pandas Datetime, the prediction will be up to that date.

Comment (JavierEscobarOrtiz) [reply skforecast/ForecasterAutoreg/ForecasterAutoreg.py]: Review all prediction methods _create_predict_inputs, predict, predict_bootstrapping, predict_interval...

**Comment by JavierEscobarOrtiz** [comment]
@imMoya incredible work! Thank you very much.

I just added some comments to keep the same format as the rest of the library.

File: skforecast/ForecasterAutoreg/ForecasterAutoreg.py (hunk 5 above)
Comment (imMoya) [reply skforecast/ForecasterAutoreg/ForecasterAutoreg.py]: @JavierEscobarOrtiz I've assumed all predict methods have the same `steps` inputs

**Review by JavierEscobarOrtiz — APPROVED** [review]

**Comment by JavierEscobarOrtiz** [comment]
@imMoya 

Thanks a lot for the PR! Lots of quality in your code.

In the next release, we will include your contribution in the release notes and your profile in the [Authors section](https://skforecast.org/latest/authors/authors).

We will keep this feature for the `ForecasterAutoreg` (which will be renamed to `ForecasterRecursive`) and, in skforecast `0.15.0` this feature will be extended to all Forecasters.

Best

**Comment by imMoya** [comment]
Thanks @JavierEscobarOrtiz for the massive support in getting me involved in this!

**Pull request merged**